fem/libceed/ceedsolvers-utility.cpp

```diff
@@ -0,0 +1,198 @@
+// Copyright (c) 2010-2020, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#include "ceedsolvers-utility.h"
+
+#include "../../general/forall.hpp"
+using namespace mfem;
+
+#ifdef MFEM_USE_CEED
+
+/// a = a (pointwise*) b
+/// @todo: using MPI_FORALL in this Ceed-like function is ugly
+int CeedVectorPointwiseMult(CeedVector a, const CeedVector b)
+{
+   int ierr;
+   Ceed ceed;
+   CeedVectorGetCeed(a, &ceed);
+
+   int length, length2;
+   ierr = CeedVectorGetLength(a, &length); CeedChk(ierr);
+   ierr = CeedVectorGetLength(b, &length2); CeedChk(ierr);
+   if (length != length2)
+   {
+      return CeedError(ceed, 1, "Vector sizes don't match");
+   }
+
+   CeedMemType mem;
+   CeedGetPreferredMemType(ceed, &mem);
+   CeedScalar *a_data;
+   const CeedScalar *b_data;
+   ierr = CeedVectorGetArray(a, mem, &a_data); CeedChk(ierr);
+   ierr = CeedVectorGetArrayRead(b, mem, &b_data); CeedChk(ierr);
+   MFEM_FORALL(i, length,
+   {a_data[i] *= b_data[i];});
```

Comment:
I don't think this is correct, I think you want to use the memory type that corresponds to MFEM (not libCEED, the `ceed` object) since you use MFEM_FORALL. So if MFEM is in device mode, then you should use `CEED_MEM_DEVICE` otherwise `CEED_MEM_HOST` for `mem`.

Reply:
This makes sense, thanks, now fixed.